File: Command/PHP/PauseCommand.php
<?php
  class PauseCommand implements Command {
    protected $player;
    protected $pre_state;

    public function __construct($player) {
      $this->player = $player;
    }

    public function excute() {
      $this->pre_state = $this->player->getState();
      $this->player->pause();
      echo $this->player->getState() . "\n";
    }

    public function undo() {
      if ($this->pre_state == "pause") {
        $this->player->pause();
      } else if ($this->pre_state == "playing") {
        $this->player->playing();
      } else if($this->pre_state == "stop") {
        $this->player->stop();
      }

      echo $this->player->getState() . "\n";
    }
  }

File: Command/Go/command/PauseCommand.go
package command

type PauseCommand struct {
  command
  *Player
  pre_state string
}

func NewPauseCommand(p *Player) *PauseCommand {
  return &PauseCommand{Player: p, pre_state: ""}
}

func (this *PauseCommand) Excute() string {
  this.pre_state = this.Player.GetState()
  this.Player.Pause()
  return this.Player.GetState()
}

func (this *PauseCommand) Undo() string {
  if this.pre_state == "pause" {
    this.Player.Pause()
  } else if this.pre_state == "playing" {
    this.Player.Playing()
  } else if this.pre_state == "stop" {
    this.Player.Stop()
  }

  return this.Player.GetState()
}
